### PR TITLE
RemoteMixtureOfExperts update part II

### PR DESCRIPTION
**Pull request by justheuristic**
* [x] RemoteExpert channels are now cached per endpoint, not per individual expert
* [x] re-write moe.py to make 1 less consecutive requests to the dht
* [x] simplify beam search - avoid unnecessary batch dimension
* [x] remove `hivemind/utils/backward.py` in favor of autograd function _RemoteCallMany
* [x] add more feature-rich tests for moe.py

Discovered but not fixed:
* background_server never shutdowns properly and always awaits a kill. 
* rapidly entering and exiting background_server multiple times without pause can hang python process

### CODE REVIEW

File: hivemind/client/moe.py [hunk 1]
```diff
@@ -99,51 +111,39 @@ def beam_search(self, grid_scores: List[torch.Tensor], k_best: int, **kwargs) ->
...
-            assert dim_scores.shape[-1] == self.grid_size[dim_index]
-
-            # create all possible successsors from current beam
-            dim_indices = np.arange(dim_scores.shape[1]).astype(str)
-            new_candidates = beam[:, :, None] + delimiters + dim_indices[None, None, :]
-            new_candidates = new_candidates.reshape([batch_size, -1])
+            # create all possible successors from current beam and sort them by total score
+            expanded_scores = beam_scores[:, None] + dim_scores[None, :]
+            sorted_indices = [(flat_i // len(dim_scores), flat_i % len(dim_scores))
+                              for flat_i in (-expanded_scores).view(-1).argsort().numpy()]
```
Comment (mryab) [inline hivemind/client/moe.py:125]: (-expanded_scores).flatten()?

File: hivemind/dht/__init__.py [hunk 2]
```diff
@@ -228,5 +230,8 @@ async def _first_k_active(
                     uid_prefixes[pre_dispatch_chunk_i * chunk_size: (pre_dispatch_chunk_i + 1) * chunk_size],
                     num_workers=num_workers_per_chunk)))
 
+        for task in pending_tasks:
+            task.cancel()
+
         # return k active prefixes or as many as we could find
-        future.set_result(active_prefixes)
+        future.set_result(OrderedDict(found))
```
[the remark this answers is not in the document]
Comment (justheuristic) [reply hivemind/dht/__init__.py:237]: Technically speaking, all dicts are ordered since python 3.7, but i want to emphasize it here for the reader.

File: tests/test_moe.py [hunk 3]
```diff
@@ -1,46 +1,127 @@
...
+        grid_scores = dmoe.proj(input).split_with_sizes(dmoe.grid_size, dim=-1)
+
+        chosen_experts = dmoe.loop.run_until_complete(dmoe.beam_search(grid_scores, k_best=dmoe.k_best))
+
+        chosen_scores = dmoe.compute_expert_scores([dim_scores[None] for dim_scores in grid_scores],
+                                                   [chosen_experts])[0]
+
+        all_scores = dmoe.compute_expert_scores([dim_scores[None] for dim_scores in grid_scores],
+                                                [[hivemind.RemoteExpert(uid, '') for uid in all_expert_uids]])[0]
+        true_best_scores = sorted(all_scores.cpu().data.numpy(), reverse=True)[:len(chosen_experts)]
```
[the remark this answers is not in the document]
Comment (justheuristic) [reply tests/test_moe.py:122]: kk, switched to .detach()

File: hivemind/client/expert.py [hunk 4]
```diff
@@ -28,22 +42,11 @@ class RemoteExpert(nn.Module):
     def __init__(self, uid, endpoint: Endpoint):
         super().__init__()
         self.uid, self.endpoint = uid, endpoint
-        self._channel, self._stub, self._info = None, None, None
+        self._info = None
```
[the remark this answers is not in the document]
Comment (justheuristic) [reply hivemind/client/expert.py:45]: Technically, we dont: other experts may still use this channel. We will benefit from _some_ closing policy or timeout, but i'm unsure as of which one it should be.

File: hivemind/client/moe.py [hunk 5]
```diff
@@ -69,25 +76,30 @@ def forward(self, input: torch.Tensor, *args: torch.Tensor, **kwargs: torch.Tens
...
-            for i in range(len(input))
-        )
+        expert_mask, *expert_outputs = _RemoteCallMany.apply(
+            DUMMY, chosen_experts, self.k_min, self.backward_k_min, self.timeout_after_k_min,
+            self.forward_timeout, self.backward_timeout, self.loop, *nested_flatten(((input, *args), kwargs)))
+        # ^-- multiple tensors of shape [batch_size, max_experts, ...output_shape]
 
-        averaged_outputs_flat = map(torch.cat, zip(*map_with_parallel_backward(_RemoteMoECall, *batch_jobs_args)))
+        expert_logits = self.compute_expert_scores(grid_scores, chosen_experts)
+        expert_logits = torch.where(expert_mask, expert_logits, torch.full_like(expert_logits, -float('inf')))
```
[the remark this answers is not in the document]
Comment (justheuristic) [reply hivemind/client/moe.py:95]: (applied both)

File: hivemind/client/moe.py [hunk 6]
```diff
@@ -69,25 +76,30 @@ def forward(self, input: torch.Tensor, *args: torch.Tensor, **kwargs: torch.Tens
...
+        expert_weights = torch.softmax(expert_logits, dim=1)
+        averaged_outputs_flat = [
+            torch.sum(expert_weights.view(*expert_weights.shape, *[1] * (tensor.ndim - 2)) * tensor, dim=1)
```
[the remark this answers is not in the document]
Comment (justheuristic) [reply hivemind/client/moe.py:98]: the problem is that you would need to unsqueeze a variable number of times
[changed it to use .flatten->multiply->view, is this better?]

File: hivemind/client/moe.py (hunk 1 above)
Comment (justheuristic) [reply hivemind/client/moe.py:125]: +

File: hivemind/dht/__init__.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (mryab) [reply hivemind/dht/__init__.py:237]: 1. The requirement of returning objects in order defined by arguments holds in many other functions (e.g. get_experts, find_nearest_nodes), but we don't hint at the order there simply because it is clear from the signature
2. Even if this case is different (not all prefixes may be returned), the way this restriction is imposed now is too implicit: the last thing the reader is going to think is "aha, we create OrderedDict, because the order of keys has to be consistent with `uid_prefixes` and not because OrderedDict has methods that `dict` doesn't / we will need order-sensitive comparison / other reasons"
3. Lastly, even with OrderedDict being created for the result, it is easy to break the assumption without anybody noticing

Thus, I suggest one of the following adjustments:
1. We can state the order preserving property in the docstring of first_k_active
2. We can avoid this assumption and restore the order in moe/beam_search (if it is really required). FYI I added `random.shuffle(beam)` after `beam = list(best_alive_prefixes.keys())` in moe.py and the tests still pass
